net/golioth/Kconfig

```diff
@@ -9,6 +9,7 @@ config GOLIOTH
 	depends on NETWORKING
 	select COAP
 	select MBEDTLS
+	select MBEDTLS_SSL_DTLS_CONNECTION_ID
```

Comment:
> enables deviecs to maintain their connection with the server for longer periods ... 

I'd like to have some evidence that backs that up, or at least theoretical use-case that enables "longer periods" of connection.

> ... and across IP address changes

What is the device use-case/scenario in which IP address changes? Isn't it like <1% of the possible cases? Personally I haven't notice IP address change at all and I can't even make up a scenario when that would happen. In case of cellular connectivity I guess that BTS roaming should solve that problem with maintaining single IP address (otherwise we would all be having connectivity problems when using mobile phones, where majority of connections/streams are TCP-based).

> resulting in lower bandwidth and reduced power usage.

I'd like to see some evidence (at least theoretical) that backs that up. I've done some tests with `qemu_x86` and `lightdb/get` sample, which shows that enabling CID support as done in this commit is **increasing** network usage (and as a result increasing power consumption). Below data shows almost 20% more data usage when sending empty CoAP messages, but there is almost 15% increase when issuing simple LightDB GET requests on `/counter` LightDB value.

<details>
<summary> No CID

`7	0.014224278	192.0.2.1	192.0.2.2	CoAP	83	NON, MID:32497, Empty Message`

</summary>

```
Frame 7: 83 bytes on wire (664 bits), 83 bytes captured (664 bits) on interface tap0, id 0
Ethernet II, Src: ICANNIAN_00:53:3d (00:00:5e:00:53:3d), Dst: 4e:a9:94:1c:22:15 (4e:a9:94:1c:22:15)
Internet Protocol Version 4, Src: 192.0.2.1, Dst: 192.0.2.2
User Datagram Protocol, Src Port: 47964, Dst Port: 5684
    Source Port: 47964
    Destination Port: 5684
    Length: 49
    Checksum: 0x8b3a [correct]
    [Checksum Status: Good]
    [Stream index: 0]
    [Timestamps]
    UDP payload (41 bytes)
Datagram Transport Layer Security
    DTLSv1.2 Record Layer: Application Data Protocol: Constrained Application Protocol
        Content Type: Application Data (23)
        Version: DTLS 1.2 (0xfefd)
        Epoch: 1
        Sequence Number: 1
        Length: 28
        Encrypted Application Data: 000100000000000120b9fceb9a7952e41b27fe19373001761d0143fe
        [Application Data Protocol: Constrained Application Protocol]
Constrained Application Protocol, Non-Confirmable, Empty Message, MID:32497
    01.. .... = Version: 1
    ..01 .... = Type: Non-Confirmable (1)
    .... 0000 = Token Length: 0
    Code: Empty Message (0)
    Message ID: 32497
```

</details>

<details>
<summary> With CID

`7	0.014537102	192.0.2.1	192.0.2.2	CoAP	103	NON, MID:44687, Empty Message`

</summary>

```
Frame 7: 103 bytes on wire (824 bits), 103 bytes captured (824 bits) on interface tap0, id 0
Ethernet II, Src: ICANNIAN_00:53:05 (00:00:5e:00:53:05), Dst: 4e:a9:94:1c:22:15 (4e:a9:94:1c:22:15)
Internet Protocol Version 4, Src: 192.0.2.1, Dst: 192.0.2.2
User Datagram Protocol, Src Port: 50091, Dst Port: 5684
    Source Port: 50091
    Destination Port: 5684
    Length: 69
    Checksum: 0x7ce1 [correct]
    [Checksum Status: Good]
    [Stream index: 0]
    [Timestamps]
    UDP payload (61 bytes)
Datagram Transport Layer Security
    DTLSv1.2 Record Layer: Application Data Protocol: Constrained Application Protocol
        Special Type: Connection ID (25)
        Version: DTLS 1.2 (0xfefd)
        Epoch: 1
        Sequence Number: 1
        Connection ID: 57fb69307e4d46b0
        Length: 40
        [Content Type: Application Data (23)]
        Encrypted Application Data: 000100000000000155fded71cbdeb98716d74510d6218ffcd2942f0ab7608e3caf2b9e18…
        [Application Data Protocol: Constrained Application Protocol]
Constrained Application Protocol, Non-Confirmable, Empty Message, MID:44687
    01.. .... = Version: 1
    ..01 .... = Type: Non-Confirmable (1)
    .... 0000 = Token Length: 0
    Code: Empty Message (0)
    Message ID: 44687
```

</details>

Reply:
> I'd like to have some evidence that backs that up, or at least theoretical use-case that enables "longer periods" of connection.

@mniestroj here is a pretty good example from Nordic: https://devzone.nordicsemi.com/nordic/nordic-blog/b/blog/posts/extend-your-battery-life-with-nrf-cloud-coap-interface#mcetoc_1habr1j851

(there is also a specific section on power-savings, but it is more focused on MQTT vs. CoAP rather than CoAP w/ CID vs. CoAP w/o CID: https://devzone.nordicsemi.com/nordic/nordic-blog/b/blog/posts/extend-your-battery-life-with-nrf-cloud-coap-interface#mcetoc_1habr3tb55)

> What is the device use-case/scenario in which IP address changes? Isn't it like <1% of the possible cases? Personally I haven't notice IP address change at all and I can't even make up a scenario when that would happen. In case of cellular connectivity I guess that BTS roaming should solve that problem with maintaining single IP address (otherwise we would all be having connectivity problems when using mobile phones, where majority of connections/streams are TCP-based).

No, I don't think it is <1% of possible cases, though it is largely dependent on the network and use case. Using a cellular network is typically going to result in more frequent IP address changes, especially if you are in transit. You can observe this just by checking your phone's IP address periodically while driving or even walking around. You can also observe it my turning off cellular connectivity (e.g. airplane mode) and turning it back on.

> I'd like to see some evidence (at least theoretical) that backs that up. I've done some tests with qemu_x86 and lightdb/get sample, which shows that enabling CID support as done in this commit is increasing network usage (and as a result increasing power consumption). Below data shows almost 20% more data usage when sending empty CoAP messages, but there is almost 15% increase when issuing simple LightDB GET requests on /counter LightDB value.

I am working on collecting some more tangible data examples, but it is fundamentally true that record size is going to increase -- we are sending more data! Therefore, the question is whether the trade-off of not having to perform handshakes as frequently outweighs that cost of sending that extra data. In some cases it may and in others it may not. However, I think it is important to factor in the cases in which power savings are the most valuable. In the case of your laptop (or more broadly, a stationary device on a WiFi network that is communicating frequently), we might see net more data sent over the network because the handshakes are already infrequent and the application data is frequent. That being said, power savings are likely comparatively less critical in that domain, and network costs are lower (e.g. WiFi vs. cellular). On the other end of the spectrum, a device that communicates over cellular, is in transit, and sleeps often is more likely to benefit from the use of connection IDs _and_ is more likely to have tighter power constraints.

Reply:
Per our discussion in office hours earlier today, we won't enable Connection ID support by default. We will write a lot of content to help clients understand when it makes sense to enable connection IDs in their own projects. I've updated this PR to remove the Kconfig changes.